src/renderers/shared/fiber/ReactFiberHostContext.js

```diff
@@ -41,54 +41,92 @@ module.exports = function<T, P, I, TI, C, CX>(
     contextDepth: number
   };
 
-  function createContainerState(rootInstance) {
-    return {
-      rootInstance,
-      contextFibers: null,
-      contextValues: null,
-      contextDepth: -1,
-    };
-  }
+  // Current container context on the stack.
+  let rootInstance : C | null = null;
+  let contextFibers : Array<Fiber | null> | null = null;
+  let contextValues : Array<CX | null> | null = null;
+  let contextDepth : number = -1;
 
-  // If we meet any portals, we'll go deeper.
-  let containerStack : Array<ContainerState> = [];
-  let containerDepth : number = -1;
+  // If we meet any portals, we'll pack outer context in this array.
+  let portalStack : Array<ContainerState> = [];
+  let portalDepth : number = -1;
 
   function getRootHostContainer() : C {
-    if (containerDepth === -1) {
-      throw new Error('Expected to find a root container.');
+    if (rootInstance == null) {
+      throw new Error('Expected root container to exist.');
     }
-    const containerState = containerStack[containerDepth];
-    const {rootInstance} = containerState;
     return rootInstance;
   }
 
-  function pushHostContainer(portalHostContainer) {
-    containerDepth++;
-    if (containerDepth === containerStack.length) {
-      containerStack[containerDepth] = createContainerState(portalHostContainer);
+  function pushPortal(portalRootInstance : C) {
+    if (rootInstance == null) {
+      throw new Error('Cannot push a portal when there is no root.');
+    }
+    // Only create a new item on the portal stack if necessary.
+    if (portalDepth === portalStack.length) {
+      portalStack[portalDepth] = {
+        rootInstance,
+        contextFibers,
+        contextValues,
+        contextDepth,
+      };
```

Comment:
Why does `context` arrays have to be stored separately here? Can't you just push another context onto the stack of contexts? Even if that is null. Then pop it on the back up.